lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,188 @@
+#include "visSharedMemWriter.hpp"
+
+#include "visSharedMemWriter.hpp"
+#include <cxxabi.h>            // for __forced_unwind
+#include <fcntl.h>             // for O_CREAT
+#include <string.h>            // for memcpy, strerror
+#include <sys/mman.h>          // for mmap, shm_open, MAP_FAILED, MAP_SHARED, PROT_READ, PROT_WRITE
+#include <sys/stat.h>          // for S_IRUSR, S_IWUSR
+#include <sys/time.h>          // for gettimeofday, timeval
+#include <sys/types.h>         // for uint
+#include <atomic>              // for atomic_bool
+#include <exception>           // for exception
+#include <functional>          // for _Bind_helper<>::type, bind, function
+#include <future>              // for async, future
+#include <map>                 // for map
+#include <regex>               // for match_results<>::_Base_type
+#include <system_error>        // for system_error
+#include <tuple>               // for get
+#include <utility>             // for pair
+#include <vector>              // for vector
+#include "Hash.hpp"            // for Hash
+#include "StageFactory.hpp"    // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "datasetManager.hpp"  // for datasetManager
+#include "datasetState.hpp"    // for freqState
+#include "fmt.hpp"             // for format, fmt
+#include "kotekanLogging.hpp"  // for INFO, DEBUG
+#include "visBuffer.hpp"       // for visFrameView, visMetadata
+#include "visUtil.hpp"         // for frameID
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+
+REGISTER_KOTEKAN_STAGE(visSharedMemWriter);
+
+visSharedMemWriter::visSharedMemWriter(Config& config, const std::string& unique_name, bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&visSharedMemWriter::main_thread, this)) {
+
+        // Fetch any simple configuration
+        root_path = config.get_default<std::string>(unique_name, "root_path", "/dev/shm/");
+        sem_name = config.get_default<std::string>(unique_name, "sem_name", "kotekan");
+        fname_access_record = config.get_default<std::string>(unique_name, "fname_access_record", "calBufferAccessRecord");
+        fname_buf = config.get_default<std::string>(unique_name, "fname_buf", "calBuffer");
+        ntime = config.get_default<size_t>(unique_name, "nsamples", 512);
+
+        // Setup the input vector
+        in_buf = get_buffer("in_buf");
+        register_consumer(in_buf, unique_name.c_str());
+
+        // Check if any of the old buffer files exist
+        DEBUG("Checking for and removing old buffer files...");
+        check_remove(root_path + "sem." + sem_name);
+        check_remove(root_path + fname_access_record);
+        check_remove(root_path + fname_buf);
+
+        // Create the semaphore, and gain first access to it
+        sem = sem_open(
```

Comment:
It might make sense to create the semaphore at the same time you create the shared mem. The state is a little confusing if one can exist without the other.